drivers/led/lp50xx.c

```diff
@@ -0,0 +1,294 @@
+{
+	return lp50xx_set_brightness(dev, led, 100);
+}
+
+static int lp50xx_off(const struct device *dev, uint32_t led)
+{
+	return lp50xx_set_brightness(dev, led, 0);
+}
+
+static int lp50xx_set_color(const struct device *dev, uint32_t led,
+			    uint8_t num_colors, const uint8_t *color)
+{
+	const struct lp50xx_config *config = dev->config;
+	const struct led_info *led_info = lp50xx_led_to_info(config, led);
+	uint8_t buf[4];
+
+	if (!led_info) {
+		return -ENODEV;
+	}
+
+	if (num_colors != led_info->num_colors) {
+		LOG_ERR("Provided number of colors is not valid");
+		return -EINVAL;
+	}
+
+	buf[0] = LP50XX_OUT_COLOR_BASE(config->chip) + 3 * led_info->index;
+	buf[1] = color[0];
+	buf[2] = color[1];
+	buf[3] = color[2];
+
+	return i2c_write_dt(&config->bus, buf, sizeof(buf));
+}
+
+static int lp50xx_write_channels(const struct device *dev,
+				 uint32_t start_channel,
+				 uint32_t num_channels, const uint8_t *buf)
+{
+	const struct lp50xx_config *config = dev->config;
+	struct lp50xx_data *data = dev->data;
+
+	if (start_channel >= LP50XX_NUM_CHANNELS(config->chip) ||
+	    start_channel + num_channels > LP50XX_NUM_CHANNELS(config->chip)) {
+		return -EINVAL;
+	}
+
+	/*
+	 * Unfortunately this controller doesn't support commands split into
+	 * two I2C messages.
+	 */
+	data->chan_buf[0] = LP50XX_CHANNEL_BASE(config->chip) + start_channel;
+	memcpy(data->chan_buf + 1, buf, num_channels);
+
+	return i2c_write_dt(&config->bus, data->chan_buf, num_channels + 1);
+}
+
+static int lp50xx_init(const struct device *dev)
+{
+	const struct lp50xx_config *config = dev->config;
+	uint8_t buf[2];
+	int err;
+
+	if (!device_is_ready(config->bus.bus)) {
+		LOG_ERR("I2C device not ready");
+		return -ENODEV;
+	}
+
+	if (config->num_leds > LP50XX_MAX_LEDS(config->chip)) {
+		LOG_ERR("%s: invalid number of LEDs %d (max %d)",
+			dev->name, config->num_leds, LP50XX_MAX_LEDS(config->chip));
+		return -EINVAL;
+	}
+
+	/* Enable EN pin if given in devicetree. */
+	if (config->en_pin.port != NULL) {
+		if (device_is_ready(config->en_pin.port)) {
+			gpio_pin_configure_dt(&config->en_pin, GPIO_OUTPUT_HIGH);
+			/* Power save mode deglitch time */
+			k_sleep(K_MSEC(40));
+		} else {
+			LOG_ERR("GPIO for EN pin is not ready!");
+			return -ENODEV;
+		}
+	}
+
+	/* Reset device */
+	buf[0] = LP50XX_DEVICE_RESET(config->chip);
+	buf[1] = LP50XX_RESET_VAL;
+	err = i2c_write_dt(&config->bus, buf, 2);
+	if (err < 0) {
+		return err;
+	}
+
+	/* Enable LED controller. */
+	buf[0] = LP50XX_DEVICE_CONFIG0;
+	buf[1] = CONFIG0_CHIP_EN;
+	err = i2c_write_dt(&config->bus, buf, 2);
+	if (err < 0) {
+		return err;
+	}
+
+	/* Apply configuration. */
+	buf[0] = LP50XX_DEVICE_CONFIG1;
+	buf[1] = CONFIG1_PWM_DITHERING_EN | CONFIG1_AUTO_INCR_EN
+		| CONFIG1_POWER_SAVE_EN;
+	if (config->max_curr_opt) {
+		buf[1] |= CONFIG1_MAX_CURRENT_OPT;
+	}
+	if (config->log_scale_en) {
+		buf[1] |= CONFIG1_LOG_SCALE_EN;
+	}
+
+	return i2c_write_dt(&config->bus, buf, 2);
```

Comment:
Same as above.

Reply:
I added these changes, but in commit with reset functionality - before this commit we have buf[3], but use only two elements in this case.